add gradscaler on CPU

CaoE opened the pull request:
Just test gradscaler on CPU




cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mcarilli @ptrblck @leslie-fang-intel

vadimkantorov commented on torch/amp/grad_scaler.py:
```diff
@@ -0,0 +1,589 @@
+                    else:
+                        to_unscale = param.grad
+
+                    # TODO: is there a way to split by device and dtype without appending in the inner loop?
```
it came up many times in `_foreach_*` methods impls, but such an util hasn't been factored out afaik :(

CaoE committed:
add gradscaler on CPU

ghstack-source-id: 5a3fadd60945a9f4b9cc2506425ddc98a033469f
Pull Request resolved: https://github.com/pytorch/pytorch/pull/102369

CaoE committed:
add test cases for gradscaler on CPU

ghstack-source-id: 9c8001d1f21aed638bbaec136352f688ee09ca13
Pull Request resolved: https://github.com/pytorch/pytorch/pull/102370

CaoE committed:
update gradscaler
